Fast Cache - Downgrade - reupgrade protection and other improvements

p0mvn opened the pull request:
**#11 - Background**

We would like to add downgrade and re-upgrade protection for the fast cache. Write the last fast node version on each block and check it at the startup. If not set to the most recent version, rewrite fast cache

**Acceptance criteria**
- [x] protection added
- [x] unit tested


**Additionally:**
- [x] optimized non-membership proof
- [x] unit tested fast node
- [x] resolved a bug related to the location of where migration was applied in `LoadVersion` and `LazyLoadVersion` 
- [x] looked into optimizing `GetWithIdndex` and `GetByIndex`
    * the only idea made it worse off so had to revert
- [x] bench tests for `GetWithIndex`, `GetByIndex` and `  


Latest benchmark for non-membership proof:

```
Running tool: /usr/local/go/bin/go test -benchmem -run=^$ -bench ^BenchmarkGetNonMembership$ github.com/cosmos/iavl

goos: linux
goarch: amd64
pkg: github.com/cosmos/iavl
cpu: Intel(R) Core(TM) i7-8650U CPU @ 1.90GHz
BenchmarkGetNonMembership/fast-8         	    1130	   1003260 ns/op	  132431 B/op	    2920 allocs/op
BenchmarkGetNonMembership/regular-8      	     154	   7399486 ns/op	 1292050 B/op	   21474 allocs/op
PASS
ok  	github.com/cosmos/iavl	54.441s
```

Old benchmark from `dev/iavl_data_locality` with modified bench tests:
```
iavl: 
git commit: 
git branch: 
go version go1.17.6 linux/amd64

Init Tree took 78.76 MB
goos: linux
goarch: amd64
pkg: github.com/cosmos/iavl/benchmarks
cpu: Intel(R) Core(TM) i7-8650U CPU @ 1.90GHz
BenchmarkMedium/memdb-100000-100-16-40/query-no-in-tree-guarantee-slow-8         	   93859	     12319 ns/op	     592 B/op	      12 allocs/op
BenchmarkMedium/memdb-100000-100-16-40/query-hits-slow-8                         	   83343	     14889 ns/op	     760 B/op	      15 allocs/op
BenchmarkMedium/memdb-100000-100-16-40/iteration-slow-8                          	       2	 829462950 ns/op	88838692 B/op	 1600081 allocs/op
--- BENCH: BenchmarkMedium/memdb-100000-100-16-40/iteration-slow-8
    bench_test.go:128: completed 100000 iterations
    bench_test.go:128: completed 100000 iterations
    bench_test.go:128: completed 100000 iterations
BenchmarkMedium/memdb-100000-100-16-40/update-8                                  	   10000	    149613 ns/op	   27337 B/op	     335 allocs/op
BenchmarkMedium/memdb-100000-100-16-40/block-8                                   	      78	  16881365 ns/op	 2910816 B/op	   35693 allocs/op
Init Tree took 40.64 MB
BenchmarkMedium/goleveldb-100000-100-16-40/query-no-in-tree-guarantee-slow-8     	   56197	     21425 ns/op	    1566 B/op	      30 allocs/op
BenchmarkMedium/goleveldb-100000-100-16-40/query-hits-slow-8                     	   43808	     27835 ns/op	    2122 B/op	      39 allocs/op
BenchmarkMedium/goleveldb-100000-100-16-40/iteration-slow-8                      	       1	1974250600 ns/op	226939520 B/op	 3862227 allocs/op
--- BENCH: BenchmarkMedium/goleveldb-100000-100-16-40/iteration-slow-8
    bench_test.go:128: completed 100000 iterations
BenchmarkMedium/goleveldb-100000-100-16-40/update-8                              	    8377	    211953 ns/op	   43961 B/op	     425 allocs/op
BenchmarkMedium/goleveldb-100000-100-16-40/block-8                               	      45	  28982847 ns/op	 5162017 B/op	   53449 allocs/op
PASS
ok  	github.com/cosmos/iavl/benchmarks	25.442s
```

Latest benchmark:
```
Running tool: /usr/local/go/bin/go test -benchmem -run=^$ -bench ^BenchmarkMedium$ github.com/cosmos/iavl/benchmarks

iavl: 
git commit: 
git branch: 
go version go1.17.6 linux/amd64

Init Tree took 114.29 MB
goos: linux
goarch: amd64
pkg: github.com/cosmos/iavl/benchmarks
cpu: Intel(R) Core(TM) i7-8650U CPU @ 1.90GHz
BenchmarkMedium/memdb-100000-100-16-40/query-no-in-tree-guarantee-fast-8         	  596952	      2612 ns/op	     112 B/op	       4 allocs/op
BenchmarkMedium/memdb-100000-100-16-40/query-no-in-tree-guarantee-slow-8         	   96513	     13278 ns/op	     440 B/op	       8 allocs/op
BenchmarkMedium/memdb-100000-100-16-40/query-hits-fast-8                         	  737870	      1686 ns/op	      26 B/op	       0 allocs/op
BenchmarkMedium/memdb-100000-100-16-40/query-hits-slow-8                         	   69254	     19333 ns/op	     605 B/op	      11 allocs/op
BenchmarkMedium/memdb-100000-100-16-40/iteration-fast-8                          	      16	 101840581 ns/op	18400314 B/op	  300000 allocs/op
--- BENCH: BenchmarkMedium/memdb-100000-100-16-40/iteration-fast-8
    bench_test.go:143: completed 100000 iterations
    bench_test.go:143: completed 100000 iterations
    bench_test.go:143: completed 100000 iterations
    bench_test.go:143: completed 100000 iterations
    bench_test.go:143: completed 100000 iterations
    bench_test.go:143: completed 100000 iterations
    bench_test.go:143: completed 100000 iterations
    bench_test.go:143: completed 100000 iterations
    bench_test.go:143: completed 100000 iterations
    bench_test.go:143: completed 100000 iterations
	... [output truncated]
BenchmarkMedium/memdb-100000-100-16-40/iteration-slow-8                          	       2	1075280050 ns/op	88843124 B/op	 1600102 allocs/op
--- BENCH: BenchmarkMedium/memdb-100000-100-16-40/iteration-slow-8
    bench_test.go:143: completed 100000 iterations
    bench_test.go:143: completed 100000 iterations
    bench_test.go:143: completed 100000 iterations
BenchmarkMedium/memdb-100000-100-16-40/update-8                                  	    8095	    226336 ns/op	   27376 B/op	     341 allocs/op
BenchmarkMedium/memdb-100000-100-16-40/block-8                                   	      52	  22879244 ns/op	 2946879 B/op	   36431 allocs/op
Init Tree took 66.82 MB
BenchmarkMedium/goleveldb-100000-100-16-40/query-no-in-tree-guarantee-fast-8     	  238384	      6958 ns/op	     814 B/op	      16 allocs/op
BenchmarkMedium/goleveldb-100000-100-16-40/query-no-in-tree-guarantee-slow-8     	   33271	     34922 ns/op	    1844 B/op	      32 allocs/op
BenchmarkMedium/goleveldb-100000-100-16-40/query-hits-fast-8                     	  459201	      2207 ns/op	     114 B/op	       2 allocs/op
BenchmarkMedium/goleveldb-100000-100-16-40/query-hits-slow-8                     	   38526	     32942 ns/op	    1987 B/op	      33 allocs/op
BenchmarkMedium/goleveldb-100000-100-16-40/iteration-fast-8                      	      10	 109896370 ns/op	29327281 B/op	  522988 allocs/op
--- BENCH: BenchmarkMedium/goleveldb-100000-100-16-40/iteration-fast-8
    bench_test.go:143: completed 100000 iterations
    bench_test.go:143: completed 100000 iterations
    bench_test.go:143: completed 100000 iterations
    bench_test.go:143: completed 100000 iterations
    bench_test.go:143: completed 100000 iterations
    bench_test.go:143: completed 100000 iterations
    bench_test.go:143: completed 100000 iterations
    bench_test.go:143: completed 100000 iterations
    bench_test.go:143: completed 100000 iterations
    bench_test.go:143: completed 100000 iterations
	... [output truncated]
BenchmarkMedium/goleveldb-100000-100-16-40/iteration-slow-8                      	       1	2197231400 ns/op	236026680 B/op	 3998799 allocs/op
--- BENCH: BenchmarkMedium/goleveldb-100000-100-16-40/iteration-slow-8
    bench_test.go:143: completed 100000 iterations
BenchmarkMedium/goleveldb-100000-100-16-40/update-8                              	    7573	    281499 ns/op	   43491 B/op	     432 allocs/op
BenchmarkMedium/goleveldb-100000-100-16-40/block-8                               	      30	  49443073 ns/op	 6370082 B/op	   66312 allocs/op
PASS
ok  	github.com/cosmos/iavl/benchmarks	49.589s
```

p0mvn committed:
add leaf hash to fast node and unit test

p0mvn committed:
refactor get with index and get by index, fix migration in load version and lazy load version

p0mvn committed:
use Get in GetVersioned of mutable tree

p0mvn committed:
refactor non membership proof to use fast storage if available

p0mvn committed:
bench non-membership proof

p0mvn committed:
fix bench tests to work with the new changes

p0mvn committed:
add downgrade-reupgrade protection and unit test

p0mvn committed:
remove leaf hash from fast node

p0mvn committed:
resolve multithreading bug related to iterators not being closed

p0mvn committed:
clean up

p0mvn committed:
use correct tree in bench tests

p0mvn committed:
add cache to tree used to bench non membership proofs

p0mvn committed:
add benc tests for GetWithIndex and GetByIndex

p0mvn commented:
It seems that GetwithIndex and GetByIndex have actually become worse off:
```
// OLD Implementation

Benchmark_GetWithIndex/regular-8      	   78734	     17354 ns/op	     720 B/op	      15 allocs/op

Benchmark_GetByIndex/regular-8      	   71974	     17146 ns/op	     822 B/op	      15 allocs/op


// NEW Implementation

Benchmark_GetWithIndex/fast-8         	      44	  23212700 ns/op	 3865367 B/op	   96616 allocs/op
Benchmark_GetWithIndex/regular-8      	   85010	     14220 ns/op	     584 B/op	      11 allocs/op

Benchmark_GetByIndex/fast-8         	     100	  20810321 ns/op	 3821985 B/op	   95532 allocs/op
Benchmark_GetByIndex/regular-8      	   72451	     17056 ns/op	     828 B/op	      15 allocs/op
```

Using fast node iterations to GetWithIndex and GetByIndex might not be the most optimal solution. It does make sense because traversing the tree allows searching for the correct key in O(log(n)) whereas iterations in O(n). The fact that O(n) fast iteration preserves data locality does not help to outperform O(log(n)) tree search with no data locality

I reverted the changes for `GetWithIndex` and `GetByIndex`.

The non-membership proof is still better off, however. Even as I was increasing the key size, the difference between old and new implementations was growing.

p0mvn committed:
revert GetWithIndex and GetByIndex

p0mvn committed:
remove unused import

p0mvn commented:
@ValarDragon Please review. This branch is diffed against my older PR #5. It also needs some more  :eyes:

ValarDragon commented on mutable_tree.go:
```diff
@@ -760,10 +761,8 @@ func (tree *MutableTree) saveFastNodeVersion() error {
 		return err
 	}
 
-	if !tree.ndb.isFastStorageEnabled() {
```
This is removed because its only called if `tree.ndb.isFastStorageEnabled` is true?

If so, can we add a comment stating this above the `saveFastNodeVersion` function?

p0mvn replied:
The reason why this was removed is that now we are going with the assumption that fast cache storage will be upgraded to on all versions above [1.1.0 or fastStorageVersionValue](https://github.com/osmosis-labs/iavl/blob/c293938ac371b008a2ac2f115b11093b49eafa5a/nodedb.go#L31). Therefore, we must make sure that we persist the latest live version on disk. Otherwise, re-upgrade protection would not function correctly.

In case there is any confusion, there are now 2 separate methods that check for fast storage to be enabled:
- in nodeDB (`hasUpgradedToFastStorage`)
  * This method checks that the upgrade has happened. When the upgrade is successful, fast storage on disk must match the latest tree.
  * I renamed this method from `isFastStorageEnabled` to `hasUpgradedToFastStorage` and added a comment for clarity 
- in immutable tree (`isFastStorageEnabled`)
   * This one checks for 2 conditions:
      1. The tree is of the latest version
      2. hasUpgradedToFastStorage - this one is more of a sanity check since now we are assuming that the upgrade should persist forever unless manually downgraded.
   * also added a comment

Let me know if this makes sense

ValarDragon commented on proof_ics23.go:
```diff
@@ -57,12 +80,57 @@ func (t *ImmutableTree) GetNonMembershipProof(key []byte) (*ics23.CommitmentProo
+	var nextKey []byte = nil
+
+	done := false
+	itr := t.Iterator(nil, nil, true)
```
Hrmm, isn't iterating over all of fast node state going to be slower than a logarithmic number of file opens? Is there a way we can set a 'smarter' start bound?

ValarDragon replied:
I think we can do this speedup in a non-db breaking way later, so may also be worth commenting out its invocation in `GetNonMembershipProof` and revisiting after we get the initial release out!

ValarDragon commented on tree_test.go:
```diff
@@ -1786,3 +1786,140 @@ func TestIterate_ImmutableTree_Version2(t *testing.T) {
+	}
+}
+
+func TestGetWithIndex_ImmutableTree(t *testing.T) {
```
Nice work with the tests!

ValarDragon commented on nodedb.go:
```diff
@@ -234,6 +250,17 @@ func (ndb *nodeDB) isFastStorageEnabled() bool {
+			return true
+		}
+	}
+	return false
```
This should never happen under correct behavior right? Should we add a comment do a fmt.Println if this occurs?

p0mvn replied:
This should happen when we have already upgraded to fast storage and it matches the latest state to avoid redundantly upgrading fast nodes when they are already live

p0mvn replied:
This should also happen if we are on the default (non-fast) storage version. This method is independent from `hasUpgradedToFastStorage`.

To avoid confusion,

hasUpgradedToFastStorage - returns true if we have upgraded to fast storage, false if we are on default

shouldForceFastStorageUpdate - returns true if we have upgraded to fast storage but it does not match the live state, false otherwise

ValarDragon commented on nodedb.go:
```diff
@@ -211,6 +219,10 @@ func (ndb *nodeDB) SaveFastNode(node *FastNode) error {
 }
 
 func (ndb *nodeDB) setStorageVersion(newVersion string) error {
+	if newVersion >= fastStorageVersionValue {
```
what does >= mean for a string here

ValarDragon replied:
Oh wow, >= is supported on strings in golang, I thought you had to use strings.Compare

Cool!

ValarDragon replied:
Can we add a comment specifying the two expected formats of each. I'm mildly worried newVersion could be called with something not complying with a semver version (e.g. with the delimeter in there)

I haven't checked all calling functions

p0mvn replied:
As per [this comment](https://github.com/osmosis-labs/iavl/pull/12#discussion_r795826447), refactored this method and addressed the issue.

ValarDragon commented on nodedb.go:
```diff
@@ -211,6 +219,10 @@ func (ndb *nodeDB) SaveFastNode(node *FastNode) error {
 
 func (ndb *nodeDB) setStorageVersion(newVersion string) error {
+	if newVersion >= fastStorageVersionValue {
+		newVersion = newVersion + fastStorageVersionDelimiter + strconv.Itoa(int(ndb.latestVersion))
```
won't we be infinitely appending to newVersion like this, if newVersion gets parsed from the serialized value later?

ValarDragon replied:
I don't think any of the tests test repeated calls of parsing & setting but maybe i'm wrong

p0mvn replied:
You are right, good catch.

I refactored this method completely and renamed it to `setFastStorageVersionToBatch`. Now, it does not take version as a parameter. Parses the `storageVersion` and updates it to fast, if needed, with the latest live tree version. Added more unit tests to make sure all code branches are covered

p0mvn committed:
unit test re-upgrade protection and fix small issues

p0mvn committed:
remove redundant setStorageVersion method

p0mvn committed:
fix bug with appending to live stage version to storage version and nit test

p0mvn committed:
add comment for setFastStorageVersionToBatch

p0mvn committed:
refactor and improve unit tests for reupgrade protection

p0mvn committed:
rename ndb's isFastStorageEnabled to hasUpgradedToFastStorage and add comments

p0mvn committed:
comment out new implementation for GetNonMembershipProof

p0mvn committed:
update comments in nodedb to reflect the difference between hasUpgradedToFastStorage and shouldForceFastStorageUpdate

p0mvn committed:
refactor nodedb tests

p0mvn commented:
@ValarDragon addressed all comments so far. The most recent updates:

- commented out `GetNonMembershipProof` new implementation for now
- addressed the bug about appending version and unit tested
- added more unit tests in mutable_tree_test.go for re-upgrade protection
- renamed some methods for clarity and added comments

Please take a look when you have time

p0mvn committed:
downgrade tendermint to 0.34.14 - osmosis's latest cosmos sdk does not support 0.35.0

p0mvn committed:
fix bug where fast storage was not enabled when version 0 was attempted to be loaded

ValarDragon commented:
Changes LGTM!!

p0mvn committed:
implement unsaved fast iterator to be used in mutable tree (#16)
